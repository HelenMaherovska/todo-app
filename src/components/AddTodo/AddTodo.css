.add-todo-wrapper {
  margin: 40px 0;
}

.add-todo-wrapper .form-control:not(:last-child) {
  border-top-right-radius: 0.25rem;
  border-bottom-right-radius: 0.25rem;
}
