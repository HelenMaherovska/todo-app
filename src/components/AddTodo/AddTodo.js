import React, { Component } from "react";
import PropTypes from "prop-types";

import { connect } from "react-redux";

import { addTodo } from "../../actions/todoActions";
import "./AddTodo.css";

export class AddTodo extends Component {
  constructor(props) {
    super(props);
    this.state = {
      inputValue: ""
    };
    this.onChange = this.onChange.bind(this);
    this.onAddTodo = this.onAddTodo.bind(this);
  }
  onChange(e) {
    this.setState({ inputValue: e.target.value });
  }

  onAddTodo() {
    const { addTodo } = this.props;
    const { inputValue } = this.state;

    this.setState({ inputValue: "" });

    return inputValue.trim() ? addTodo(inputValue) : null;
  }

  render() {
    const { inputValue } = this.state;
    return (
      <div className="add-todo-wrapper">
        <div className="input-group d-flex pr-0 pl-0 justify-content-between">
          <input
            type="text"
            className="form-control mr-4"
            value={inputValue}
            onChange={this.onChange}
          />
          <button
            type="submit"
            className="btn btn-success"
            onClick={this.onAddTodo}
          >
            Add Todo
          </button>
        </div>
      </div>
    );
  }
}

AddTodo.propTypes = {
  addTodo: PropTypes.func.isRequired
};

const mapDispatchToProps = {
  addTodo
};

export default connect(
  null,
  mapDispatchToProps
)(AddTodo);
