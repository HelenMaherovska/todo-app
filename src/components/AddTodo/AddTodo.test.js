import React from "react";
import { shallow } from "enzyme";
import { AddTodo } from "./AddTodo";

describe("AddTodo", () => {
  const text = "Learn unit testing";
  const mockAddTodo = jest.fn();
  const props = { todo: { text }, addTodo: mockAddTodo };
  const addTodo = shallow(<AddTodo {...props} />);

  it("renders correctly", () => {
    expect(addTodo).toMatchSnapshot();
  });
  it("should have the `Add Todo` button", () => {
    expect(addTodo.find(".btn-success").text()).toEqual("Add Todo");
  });

  it("initilizes an inputValue in `state`", () => {
    expect(addTodo.state()).toEqual({ inputValue: "" });
  });

  describe("before typing into the add todo input", () => {
    const inputValue = "Learn unit testing";

    beforeEach(() => {
      addTodo
        .find(".form-control")
        .simulate("change", { target: { value: inputValue } });
    });

    it("updates the inputValue in `state`", () => {
      expect(addTodo.state().inputValue).toEqual(inputValue);
    });
  });

  describe("when clicking the `Add Todo` button", () => {
    beforeEach(() => {
      addTodo.find(".btn-success").simulate("click");
    });

    it("successfully calls the onClick handler", () => {
      expect(mockAddTodo.mock.calls.length).toEqual(1);
    });

    it("dispatches the `addTodo()` it receives from props", () => {
      expect(mockAddTodo).toHaveBeenCalledWith(text);
    });
  });

  describe("after clicking the `Add Todo` button", () => {
    it("inputValue in `state` should be empty", () => {
      expect(addTodo.state()).toEqual({ inputValue: "" });
    });
  });
});
