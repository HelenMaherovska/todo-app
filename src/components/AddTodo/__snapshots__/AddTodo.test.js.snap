// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTodo renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddTodo
    addTodo={[MockFunction]}
    todo={
      Object {
        "text": "Learn unit testing",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="input-group d-flex pr-0 pl-0 justify-content-between"
      >
        <input
          className="form-control mr-4"
          onChange={[Function]}
          type="text"
          value=""
        />
        <button
          className="btn btn-success"
          onClick={[Function]}
          type="submit"
        >
          Add Todo
        </button>
      </div>,
      "className": "add-todo-wrapper",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            className="form-control mr-4"
            onChange={[Function]}
            type="text"
            value=""
          />,
          <button
            className="btn btn-success"
            onClick={[Function]}
            type="submit"
          >
            Add Todo
          </button>,
        ],
        "className": "input-group d-flex pr-0 pl-0 justify-content-between",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "form-control mr-4",
            "onChange": [Function],
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Add Todo",
            "className": "btn btn-success",
            "onClick": [Function],
            "type": "submit",
          },
          "ref": null,
          "rendered": "Add Todo",
          "type": "button",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="input-group d-flex pr-0 pl-0 justify-content-between"
        >
          <input
            className="form-control mr-4"
            onChange={[Function]}
            type="text"
            value=""
          />
          <button
            className="btn btn-success"
            onClick={[Function]}
            type="submit"
          >
            Add Todo
          </button>
        </div>,
        "className": "add-todo-wrapper",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="form-control mr-4"
              onChange={[Function]}
              type="text"
              value=""
            />,
            <button
              className="btn btn-success"
              onClick={[Function]}
              type="submit"
            >
              Add Todo
            </button>,
          ],
          "className": "input-group d-flex pr-0 pl-0 justify-content-between",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "form-control mr-4",
              "onChange": [Function],
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Add Todo",
              "className": "btn btn-success",
              "onClick": [Function],
              "type": "submit",
            },
            "ref": null,
            "rendered": "Add Todo",
            "type": "button",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
